refgenconf/refgenconf.py

```diff
@@ -2087,11 +2166,155 @@ def _select_genomes(self, genome=None, strict=False, order=None,
                 else:
                     missing.append(g)
             if missing:
-                _LOGGER.warning("Genomes do not include: {}".format(", ".join(missing)))
+                _LOGGER.warning(
+                    "Genomes do not include: {}".format(", ".join(missing)))
             return None if not filtered else filtered
         return genomes if not all(x in genomes for x in genome) else genome
 
 
+def config_upgrade(target_version, filepath, force=False, get_json_url=lambda server: construct_request_url(server, API_ID_ALIAS_DIGEST), link_fun=lambda s, t: os.symlink(s, t)):
```

Comment:
`target_version` argument is not used anywhere apart from messaging. If I requested an upgrade to 0.5, this function would disregard that and upgrade to 0.4.

We need to keep a mapping of possible conversions in `const` module. For now we will have just 0.3->0.4, but in the future other conversions may be available. Importantly, if I request unavailable one it needs raise an exception, a `ValueError` for example.

Reply:
should we also take a `current_version` param for this function? is it possible we will have 0.4->0.5? in that case `rgc = _RefGenConfV03(filepath=filepath, writable=True)` won't work.  and we can't use `rgc[CFG_VERSION_KEY]` because the `rgc` is not created yet... Or are we only considering one version behind, where we just hard code `rgc = _RefGenConfV{current_version}(filepath=filepath, writable=True)` every time the new release requires config upgrade?

Reply:
> should we also take a current_version param for this function?

It should get this from the file, this piece of information is available there.

> is it possible we will have 0.4->0.5? 

sure, there will be such conversion in the future.

> Or are we only considering one version behind, where we just hard code rgc = _RefGenConfV{current_version}(filepath=filepath, writable=True) every time the new release requires config upgrade?

There's more than one way to do this, but to make it simple let's just divide the function into conditionals, one per conversion type. So will have just one for this release. The "Avail conversions mapping" I mentioned before will help you structure this and react to infeasible conversions.